### DISCUSSION TITLE
Port StoryBuilder to UNO Framework

### DISCUSSION THREAD

**Discussion by terrycox** [storybuilder-org/StoryCAD#274] [Ideas]
In order to attract as wide a user base as possible, StoryBuilder must run on as many devices as possible. Currently, WinUI 3 only works on Windows Platforms. 

According to Wikipedia, for desktop and laptop computers, Windows is the most used at 75%, followed by Apple's [macOS] at 16%, and Linux-based operating systems, including Google's Chrome OS, at 5% (thereof " desktop Linux " at 2.35%). With tablets, Apple's iOS has 55% and Android has 45%.

One way StoryCAD can run on more devices is to port the app to UNO.
Another would be through MAUI.
And yet another option is to consider Qt#.

This discussion started as issue #269.
See also:
https://platform.uno/docs/articles/uwp-vs-winui3.html?msclkid=f43a1190b58311ecb66a53d1aff6eb21

**Comment by Rarisma** [comment]
#StoryCAD booting and editing stories

Auto/Releasebuilder would require updating to build UNO projects

Replace System.Drawing in NRTF
Apparently has some compatibility issues on non windows platforms, MS has a page with some dropin replacements

Remove WinUIEx and CommunityToolkit.WinUI.UI.Controls.Core
Shouldn't be much of an issue as they aren't used widely in StoryCAD

Replace MySQL Package with MySqlConnector
Non-Windows compatability, drop in replacement just need to adjust namspaces and member names

Some classes require partial adding to them
Simple, but required to compile

Printing implementation needs be changed/modified
Current printing implementation isn't compatible

Single Instancing needs to be changed/removed
Current implementation isn't compatible and doesn't make sense on platforms like iOS/Android/WASM 

File/Folder Openeing needs to be tweaked
Small tweaks with HWIND initalisation for it work.

Tweaks in App.xaml.cs to work with Uno
UNO initalises differently to WinUI so its just a bit of a porting job

Manifest Resources must be placed at the root of a project instead of in a folder, Im reporting this to the UNO Team.

Per platform incompatabilities
Some things can be incompatible

Certain future projects might have issues porting to uno

**Comment by Rarisma** [comment]
We have decided post 3.1 to investigate porting StoryCAD to UNO, multiple sub-issues will appear regarding this.
Currently thoughf ro documentation purposes the following has been done within UNOTestBranch to make StoryCAD and StoryCADLib build for UNO Windows and UNO Desktop (Linux).

- WinUIEX has been removed
- Launching via a file has been removed
- Single Instancing has been removed
- Navigation has been reworked to not need bindable page, and bindable page has been removed
- All pages that relied on bindable page have been updated to be the standard page.
- Some StoryCAD Collaborator (Unreleased) code has been commented out due to compile issues
- Internal assembly namespaces for StoryCADLib have been updated to StoryCADLib (used in ManifestResource
- Printing has been disabled due to compatability issues
- XAML binds have been updated to remove trailing spaces after , as this breaks builds on non windows platforms
- Right Click menu has been disabled on nonwindows platforms due to the current form not being implemented.
- The file availablity check (checks if a file is local or on onedrive) has been disabled on nonwindows platforms.
- OpenFile/OpenFolder in windowing have been tweaked to not log file attributes on nonwindows platforms.

**Comment by Rarisma** [comment]
Complete in 4.0.
